OpenRA.Game/FileSystem/BagFile.cs

```diff
@@ -0,0 +1,206 @@
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("WAVE"));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("fmt "));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(20));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)WavLoader.WaveType.ImaAdpcm));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)channels));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(entry.SampleRate));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(bytesPerSec));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)chunkSize));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)4));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)2));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)samplesPerChunk));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("fact"));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(4));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(4 * entry.Length));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("data"));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(entry.Length));
+			}
+
+			waveHeaderMemoryStream.Seek(0, SeekOrigin.Begin);
+
+			// Construct a merged stream
+			var mergedStream = new MergedStream(waveHeaderMemoryStream, s);
+			mergedStream.SetLength(waveHeaderMemoryStream.Length + entry.Length);
+
+			return mergedStream;
+		}
+
+		uint? FindMatchingHash(string filename)
+		{
+			var hash = IdxEntry.HashFilename(filename, type);
+			if (index.ContainsKey(hash))
+				return hash;
+
+			// Maybe we were given a raw hash?
+			uint raw;
+			if (!uint.TryParse(filename, NumberStyles.AllowHexSpecifier, CultureInfo.InvariantCulture, out raw))
+				return null;
+
+			if ("{0:X}".F(raw) == filename && index.ContainsKey(raw))
+				return raw;
+
+			return null;
+		}
+
+		public Stream GetContent(string filename)
+		{
+			var hash = FindMatchingHash(filename);
+			return hash.HasValue ? GetContent(hash.Value) : null;
+		}
+
+		public bool Exists(string filename)
+		{
+			return FindMatchingHash(filename).HasValue;
+		}
+
+		public IEnumerable<uint> ClassicHashes()
+		{
+			return Nothing;
+		}
+
+		public IEnumerable<uint> CrcHashes()
+		{
+			return index.Keys;
+		}
+
+		public IEnumerable<string> AllFileNames()
+		{
+			var lookup = new Dictionary<uint, string>();
+			if (Exists("local mix database.dat"))
+			{
+				var db = new XccLocalDatabase(GetContent("local mix database.dat"));
+				foreach (var e in db.Entries)
+				{
+					var hash = IdxEntry.HashFilename(e, type);
+					if (!lookup.ContainsKey(hash))
+						lookup.Add(hash, e);
+				}
+			}
+
+			if (GlobalFileSystem.Exists("global mix database.dat"))
+			{
+				var db = new XccGlobalDatabase(GlobalFileSystem.Open("global mix database.dat"));
+				foreach (var e in db.Entries)
+				{
+					var hash = IdxEntry.HashFilename(e, type);
+					if (!lookup.ContainsKey(hash))
+						lookup.Add(hash, e);
```

Comment:
The duplication with the code a few lines above is avoidable if you place the file exists next to each other and set `XccGlobalDatabase` accordingly.